system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,222 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
+
+  G_LOCK (ongoing_pulls);
+  g_ptr_array_remove (ongoing_pulls, pull);
+  G_UNLOCK (ongoing_pulls);
+
+  g_slice_free (FlatpakSystemHelperOngoingPull, pull);
+}
+
+static void
+system_helper_ongoing_pull_teardown (FlatpakSystemHelperOngoingPull *pull)
+{
+  if (pull->bindfs != NULL)
+    {
+      /* We are probably here because graceful unmounting did not happen during handle_deploy.
+         Hence, we try force unmount our bindfs directory, to make sure no open FDs can write
+         to it anymore. If this fails, we kill the subprocess and try to cleanup /etc/mtab
+         mount entry. */
+      if (umount2 (pull->user_dir_name, MNT_FORCE) == -1)
+        {
+          g_autoptr(GError) local_error = NULL;
+          glnx_throw_errno_prefix (&local_error, "Failed to unmount bindfs dir %s", pull->user_dir_name);
+          g_warning ("%s; killing subprocess", local_error->message);
+          g_clear_error (&local_error);
+
+          g_subprocess_force_exit (pull->bindfs);
+          /* Perform a lazy unmount operation to cleanup the dangling mount entry in /etc/mtab due to SIGKILL. */
+          if (umount2 (pull->user_dir_name, MNT_DETACH) == -1)
+            {
+              glnx_throw_errno_prefix (&local_error, "Failed to umount2 bindfs dir %s", pull->user_dir_name);
+              g_warning ("%s; mount maybe dangling in /etc/mtab", local_error->message);
+            }
+        }
+    }
+
+  if (pull->watch_id)
+    g_clear_handle_id (&pull->watch_id, g_source_remove);
+
+  if (pull->unsuccessful_mount_timeout_id)
+    g_clear_handle_id (&pull->unsuccessful_mount_timeout_id, g_source_remove);
+
+  if (pull->mount_monitor_id)
+    {
+      g_signal_handler_disconnect (pull->mount_monitor, pull->mount_monitor_id);
+      pull->mount_monitor_id = 0;
+    }
+}
+
+static void
+system_helper_ongoing_pull_unref (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_return_if_fail (pull != NULL);
+  g_return_if_fail (pull->ref_count > 0);
+
+  if (g_atomic_int_dec_and_test (&pull->ref_count))
+    {
+      system_helper_ongoing_pull_teardown (pull);
+      system_helper_ongoing_pull_free (pull);
+    }
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (FlatpakSystemHelperOngoingPull, system_helper_ongoing_pull_unref);
+
+static FlatpakSystemHelperOngoingPull *
+system_helper_ongoing_pull_ref (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_return_val_if_fail (pull != NULL, NULL);
+  g_return_val_if_fail (pull->ref_count > 0, NULL);
+
+  g_atomic_int_inc (&pull->ref_count);
+  return pull;
+}
+
+static void
+name_vanished_callback (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  g_autoptr(FlatpakSystemHelperOngoingPull) pull = user_data;
+}
+
+static void
+mounts_changed (GUnixMountMonitor *monitor, gpointer user_data)
+{
+  g_autoptr(FlatpakSystemHelperOngoingPull) pull = user_data;
+  g_autoptr(GList) mounts = NULL;
+  GUnixMountEntry *mount_entry = NULL;
+  GList *l = NULL;
+  gboolean mount_successful = FALSE;
+
```

Comment:
this could conceivably be triggered multiple times ? Should probably check if pull->mount_monitor_id is already 0 and bail out ?

Reply:
Well, sure it can trigger multiple times, but if we found our mount is successful, we disconnect this handler immediately. See below in this same function.

Reply:
lets say its fine